### DISCUSSION TITLE
How do I incorporate a scheduler with "step" and "batch step"?

### DISCUSSION THREAD

**Discussion by shreyaskamathkm** [Lightning-AI/pytorch-lightning#6234] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi,

I am trying to incorporate the scheduler provided [here](https://github.com/mpyrozhok/adamwr). 

I went through the documentation of PyTorch lightning and it allows only scheduler interval "epoch" or "step". But in the scheduler attached, it needs to call "step" every epoch and  "batch step" after each iteration. Could someone help me figure out a way to do this? Any help will be appreciated.

Thank you,
Best,
Shreyas Kamath

**Comment by SkafteNicki** [comment, marked as answer]
It is currently not supported that a scheduler can step on both batch and epoch. The scheduler you are linking to are also different from a standard scheduler since it both has an `step` method and a `batch_step` method, where lightning only supports the `step` method. 
To get this working in lightning, you would need to split the scheduler into two new schedulers that both operate on the same optimizer: one that steps on batch and one the steps on epoch.

**Reply by Alex17Li** [reply]
How can one create two schedulers that operate on the same optimizer? https://github.com/Lightning-AI/lightning/discussions/8882 suggests that the way that I would expect it to work does not work
